File: Software/LiDARConnection.py
from socket import timeout
import serial
import time

#Serial port variables
#Permission problems under Linus use: sudo chmod 666 /dev/ttyUSB0 
SERIAL_PORT_WIN = "COM3"
SERIAL_PORT_LINUX = '/dev/ttyUSB0'
SERIAL_PORT = SERIAL_PORT_LINUX
SERIAL_BAUDRATE = 115200

#Scan variables
scanSamplesSignalQuality= [0.0]
scanSamplesRange = [0.0]

#Delta-2G Frame Characteristics
#Constant Frame Parts values
FRAME_HEADER = 0xAA			#Frame Header value
PROTOCOL_VERSION = 0x01		#Protocol Version value
FRAME_TYPE = 0x61			#Frame Type value
#Scan Characteristics
SCAN_STEPS = 15				#How many steps/frames each full 360deg scan is composed of
#Received value scaling
ROTATION_SPEED_SCALE = 0.05 * 60	#Convert received value to RPM (LSB: 0.05 rps)
ANGLE_SCALE = 0.01			#Convert received value to degrees (LSB: 0.01 degrees)
RANGE_SCALE = 0.25 * 0.001	#Convert received value to meters (LSB: 0.25 mm)

#Delta-2G frame structure
class Delta2GFrame:
	frameHeader = 0			#Frame Header: 1 byte
	frameLength = 0			#Frame Length: 2 bytes, from frame header to checksum (excluded)
	protocolVersion = 0		#Protocol Version: 1 byte
	frameType = 0			#Frame Type: 1 byte
	commandWord = 0			#Command Word: 1 byte, identifier to distinguish parameters
	parameterLength = 0		#Parameter Length: 2 bytes, length of the parameter field
	parameters = [0]		#Parameter Field
	checksum = 0			#Checksum: 2 bytes

def LiDARFrameProcessing(frame: Delta2GFrame):
	match frame.commandWord:
		case 0xAE:
			#Device Health Information: Speed Failure
			rpm = frame.parameters[0] * ROTATION_SPEED_SCALE
			print("RPM: %f" % rpm)
		case 0xAD:
			#1st: Rotation speed (1 byte)
			rpm = frame.parameters[0] * ROTATION_SPEED_SCALE
			print("RPM: %f" % rpm)

			#2nd: Zero Offset angle (2 bytes)
			offsetAngle = (frame.parameters[1] << 8) + frame.parameters[2]
			offsetAngle = offsetAngle * ANGLE_SCALE

			#3rd: Start angle of current data freame (2 bytes)
			startAngle = (frame.parameters[3] << 8) + frame.parameters[4]
			startAngle = startAngle * ANGLE_SCALE

			#Calculate number of samples in current frame
			sampleCnt = int((frame.parameterLength - 5) / 3)

			#Calculate current angle index of a full frame: For Delta-2G each full rotation has 15 frames
			frameIndex = int(startAngle / (360.0 / SCAN_STEPS))

			if frameIndex == 0:
				#New scan started
				scanSamplesRange.clear()
				scanSamplesSignalQuality.clear()

			#4th: LiDAR samples, each sample has: Signal Value/Quality (1 byte), Distance Value (2 bytes)
			for i in range(sampleCnt):
				signalQuality = frame.parameters[5 + (i * 3)]
				distance = (frame.parameters[5 + (i * 3) + 1] << 8) + frame.parameters[5 + (i * 3) + 2]
				scanSamplesSignalQuality.append(signalQuality)
				scanSamplesRange.append(distance * RANGE_SCALE)

			if frameIndex == (SCAN_STEPS - 1):
				#Scan complete
				print("SCAN: %d" % len(scanSamplesRange))

def main():
	try:
		lidarSerial = serial.Serial(SERIAL_PORT, SERIAL_BAUDRATE, timeout=0)
	except serial.serialutil.SerialException:
		print("ERROR: Serial Connection Error")
		return

	status = 0
	checksum = 0
	lidarFrame = Delta2GFrame()
	while True:
		rx = lidarSerial.read(100)

		for by in rx:
			match status:
				case 0:
					#1st frame byte: Frame Header
					lidarFrame.frameHeader = by
					if lidarFrame.frameHeader == FRAME_HEADER:
						#Valid Header
						status = 1
					else:
						print("ERROR: Frame Header Failed")
					#Reset checksum, new frame start
					checksum = 0
				case 1:
					#2nd frame byte: Frame Length MSB
					lidarFrame.frameLength = (by << 8)
					status = 2
				case 2:
					#3rd frame byte: Frame Length LSB
					lidarFrame.frameLength += by
					status = 3
				case 3:
					#4th frame byte: Protocol Version
					lidarFrame.protocolVersion = by
					if lidarFrame.protocolVersion == PROTOCOL_VERSION:
						#Valid Protocol Version
						status = 4
					else:
						print("ERROR: Frame Protocol Version Failed")
						status = 0
				case 4:
					#5th frame byte: Frame Type
					lidarFrame.frameType = by
					if lidarFrame.frameType == FRAME_TYPE:
						#Valid Frame Type
						status = 5
					else:
						print("ERROR: Frame Type Failed")
						status = 0
				case 5:
					#6th frame byte: Command Word
					lidarFrame.commandWord = by
					status = 6
				case 6:
					#7th frame byte: Parameter Length MSB
					lidarFrame.parameterLength = (by << 8)
					status = 7
				case 7:
					#8th frame byte: Parameter Length LSB
					lidarFrame.parameterLength += by
					lidarFrame.parameters.clear()
					status = 8
				case 8:
					#9th+ frame bytes: Parameters
					lidarFrame.parameters.append(by)
					if len(lidarFrame.parameters) == lidarFrame.parameterLength:
						#End of parameter frame bytes
						status = 9
				case 9:
					#N+1 frame byte: Checksum MSB
					lidarFrame.checksum = (by << 8)
					status = 10
				case 10:
					#N+2 frame byte: Checksum LSB
					lidarFrame.checksum += by
					#End of frame reached
					#Compare received and calculated frame checksum
					if lidarFrame.checksum == checksum:
						#Checksum match: Valid frame
						LiDARFrameProcessing(lidarFrame)
					else:
						#Checksum missmatach: Invalid frame
						print("ERROR: Frame Checksum Failed");
					status = 0
			#Calculate current frame checksum, all bytes excluding the last 2, which are the checksum
			if status < 10:
				checksum = (checksum + by) % 0xFFFF

if __name__ == "__main__":
	main()

File: Software/LiDARConnection.cs
using System;
using System.IO.Ports;
using System.Threading;

public class LiDARConnection {
	//Serial Port Variables
	string port = "COM3";
	int baudrate = 115200;
	bool rxUSBThreadRun = true;
	Thread rxUSBThread;
	SerialPort serialPort = new SerialPort();

	//Scan variables
	int scanSamplesIndex;
	double[] scanSamplesSignalQuality;
	double[] scanSamplesRange;

	//Delta-2G Frame Structure
	struct Delta2GFrame {
		public int frameHeader;			//Frame Header: 1 byte
		public int frameLength;			//Frame Length: 2 bytes, from frame header to checksum (excluded)
		public int protocolVersion;		//Protocol Version: 1 byte
		public int frameType;			//Frame Type: 1 byte
		public int commandWord;			//Command Word: 1 byte, identifier to distinguish parameters
		public int parameterLength;		//Parameter Length: 2 bytes, length of the parameter field
		public byte[] parameters;		//Parameter Field
		public int checksum;			//Checksum: 2 bytes
	}

	//Delta-2G Frame Characteristics
	//Constant Frame Parts values
	const int frameHeader = 0xAA;                       //Frame Header value
	const int protocolVersion = 0x01;                   //Protocol Version value
	const int frameType = 0x61;							//Frame Type value
	//Scan Characteristics
	const int scanSteps = 15;							//How many steps/frames each full 360deg scan is composed of
	//Received value scaling
	const double rotationSpeedScale = 0.05 * 60;        //Convert received value to RPM (LSB: 0.05 rps)
	const double angleScale = 0.01;                     //Convert received value to degrees (LSB: 0.01 degrees)
	const double rangeScale = 0.25 * 0.001;				//Convert received value to meters (LSB: 0.25 mm)

	Action<double[], double[], int> onNewScanReceiveCallback;
	Action<int> onNewRPMReceiveCallback;

	public LiDARConnection() {

	}

	~LiDARConnection() {
		this.Disconnect();
	}

	public bool Connect(string port, int baudrate, Action<double[], double[], int> scanCallback, Action<int> rpmCallback) {
		if (serialPort.IsOpen == false) {
			try {
				serialPort.PortName = port;
				serialPort.BaudRate = baudrate;
				serialPort.DataBits = 8;
				serialPort.StopBits = StopBits.One;
				serialPort.Parity = Parity.None;
				serialPort.NewLine = "\n";
				serialPort.Open();

				this.port = port;
				this.baudrate = baudrate;
				this.onNewScanReceiveCallback = scanCallback;
				this.onNewRPMReceiveCallback = rpmCallback;

				rxUSBThreadRun = true;
				rxUSBThread = new Thread(USBRXThread);
				rxUSBThread.Start();
			}
			catch (System.Exception ex) {
				Console.WriteLine("ERROR: Serial Connect Error" + ex.Message);
				return false;
			}
		}
		else {
			return false;
		}
		return true;
	}

	public bool Disconnect() {
		if (serialPort.IsOpen == true) {
			//Tell RX thread to stop
			rxUSBThreadRun = false;
			//Wait for thread to stop
			rxUSBThread.Join();
			//Close serial port
			serialPort.Close();
		}
		else {
			return false;
		}
		return true;
	}

	public bool IsConnected() {
		return serialPort.IsOpen;
	}

	private void USBRXThread() {
		UInt16 checksum = 0;
		int status = 0;
		int index = 0;
		Delta2GFrame lidarFrame = new Delta2GFrame() { frameLength = 0, parameterLength = 0};

		byte[] rxBuffer = new byte[2048];
		while (rxUSBThreadRun) {
			if (serialPort.IsOpen) {
				int rxLength = 0;
				try {
					rxLength = serialPort.BaseStream.Read(rxBuffer, 0, rxBuffer.Length);
				}
				catch {
					Console.WriteLine("ERROR: serialPort.BaseStream.Read()");
					return;
				}

				for (int i = 0; i < rxLength; i++) {
					switch (status) {
						case 0:
							//1st frame byte: Frame Header
							lidarFrame.frameHeader = rxBuffer[i];
							if(lidarFrame.frameHeader == frameHeader) {
								//Valid Header
								status = 1;
							}
							else {
								Console.WriteLine("ERROR: Frame Header Failed");
							}
							//Reset checksum, new frame start
							checksum = 0;
							break;
						case 1:
							//2nd frame byte: Frame Length MSB
							lidarFrame.frameLength = (rxBuffer[i] << 8);
							status = 2;
							break;
						case 2:
							//3rd frame byte: Frame Length LSB
							lidarFrame.frameLength += rxBuffer[i];
							status = 3;
							break;
						case 3:
							//4th frame byte: Protocol Version
							lidarFrame.protocolVersion = rxBuffer[i];
							if(lidarFrame.protocolVersion == protocolVersion) {
								//Valid Protocol Version
								status = 4;
							}
							else {
								Console.WriteLine("ERROR: Frame Protocol Version Failed");
								status = 0;
							}
							break;
						case 4:
							//5th frame byte: Frame Type
							lidarFrame.frameType = rxBuffer[i];
							if (lidarFrame.frameType == frameType) {
								//Valid Frame Type
								status = 5;
							}
							else {
								Console.WriteLine("ERROR: Frame Type Failed");
								status = 0;
							}
							break;
						case 5:
							//6th frame byte: Command Word
							lidarFrame.commandWord = rxBuffer[i];
							status = 6;
							break;
						case 6:
							//7th frame byte: Parameter Length MSB
							lidarFrame.parameterLength = (rxBuffer[i] << 8);
							status = 7;
							break;
						case 7:
							//8th frame byte: Parameter Length LSB
							lidarFrame.parameterLength += rxBuffer[i];
							lidarFrame.parameters = new byte[lidarFrame.parameterLength];
							index = 0;
							status = 8;
							break;
						case 8:
							//9th+ frame bytes: Parameters
							lidarFrame.parameters[index++] = rxBuffer[i];
							if(index == lidarFrame.parameterLength) {
								//End of parameter frame bytes
								status = 9;
							}
							break;
						case 9:
							//N+1 frame byte: Checksum MSB
							lidarFrame.checksum = (rxBuffer[i] << 8);
							status = 10;
							break;
						case 10:
							//N+2 frame byte: Checksum LSB
							lidarFrame.checksum += rxBuffer[i];
							//End of frame reached
							//Compare received and calculated frame checksum
							if(lidarFrame.checksum == checksum) {
								//Checksum match: Valid frame
								LiDARFrameProcessing(lidarFrame);
							}
							else {
								//Checksum missmatach: Invalid frame
								Console.WriteLine("ERROR: Frame Checksum Failed");
							}
							status = 0;
							break;
						default:
							status = 0;
							break;
					}

					//Calculate current frame checksum, all bytes excluding the last 2, which are the checksum
					if(status < 10) {
						checksum += rxBuffer[i];
					}
				}
			}
		}
	}

	private void LiDARFrameProcessing(Delta2GFrame frame) {
		int cmdCode = frame.commandWord;
		switch (cmdCode) {
			case 0xAE: {
				//Device Health Information: Speed Failure
				int rpm = (int)(frame.parameters[0] * rotationSpeedScale);
				onNewRPMReceiveCallback(rpm);
				break;
			}
			case 0xAD: {
				//1st: Rotation speed (1 byte)
				int rpm = (int)(frame.parameters[0] * rotationSpeedScale);
				onNewRPMReceiveCallback(rpm);

				//2nd: Zero Offset angle (2 bytes)
				int offsetAngleInt = (frame.parameters[1] << 8) + (frame.parameters[2]);
				double offsetAngle = offsetAngleInt * angleScale;

				//3rd: Start angle of current data freame (2 bytes)
				int startAngleInt = (frame.parameters[3] << 8) + frame.parameters[4];
				double startAngle = startAngleInt * angleScale;

				//Calculate number of samples in current frame
				int samplesCnt = (frame.parameterLength - 5) / 3;

				//Calculate current angle index of a full frame: For Delta-2G each full rotation has 15 frames
				int frameIndex = (int)(startAngle / (360.0 / scanSteps));

				if (frameIndex == 0) {
					//New scan started
					scanSamplesRange = new double[samplesCnt * scanSteps];
					scanSamplesSignalQuality = new double[samplesCnt * scanSteps];
					scanSamplesIndex = 0;
				}

				//4th: LiDAR samples, each sample has: Signal Value/Quality (1 byte), Distance Value (2 bytes)
				for (int i = 0; i < samplesCnt; i++) {
					int signalQuality = frame.parameters[5 + (i * 3)];
					int distance = (frame.parameters[5 + (i * 3) + 1] << 8) + frame.parameters[5 + (i * 3) + 2];
					if (scanSamplesRange != null && scanSamplesIndex < scanSamplesRange.Length) {
						scanSamplesSignalQuality[scanSamplesIndex] = signalQuality;
						scanSamplesRange[scanSamplesIndex++] = distance * rangeScale;
					}
				}

				if (frameIndex == (scanSteps - 1)) {
					//Scan complete
					if (scanSamplesRange != null && scanSamplesIndex == scanSamplesRange.Length) {
						//Valid scan array
						onNewScanReceiveCallback(scanSamplesRange, scanSamplesSignalQuality, scanSamplesRange.Length);
					}
				}
				break;
			}
		}
	}
}
